### DISCUSSION TITLE
Hot Module Replacement in VS code Webview with React

### DISCUSSION THREAD

**Discussion by umar-shabbir** [microsoft/vscode-discussions#739] [Extension Development QnA] [answered]
How to achieve Hot Module Replacement for VS code Webview UI implemented using React? Just like normal React apps that we run in a browser, it support reloading changes without browser refresh. This is not the case with Extension Development Host window of VS code. Is there any way to achieve this for previewing UI changes in a VS code extension host window without reloading the Webview manually?

**Comment by isidorn** [comment]
Did you try something, and how did it not work exactly?

**Reply by umar-shabbir** [reply]
I could not find any relevant material to try out from the VS code extension guides. I am new to React so do not know much about it. I used this sample to get started [React + Vite](https://github.com/microsoft/vscode-webview-ui-toolkit-samples/tree/main/frameworks/hello-world-react-vite)
For extension logic I am running typescript compiler in watch mode and for the react webview ui I am using "vite build --watch"
This way, any changes to the source are compiled when I save them. But the changes does not show up in the Extension Development Host window after compilation. I have to reload this window to view my changes to the web view

**Reply by isidorn** [reply]
I suggest to read this https://code.visualstudio.com/api/extension-capabilities/overview
I am not sure if using React is the way to go here. Since VS Code extensions can not contribute a lot of UI. Extensions are more like data providers.

**Reply by umar-shabbir** [reply]
I have gone through this document. This is not relevant as it is more about what extensions can do. 
My specific extension is targeting the Webview based extension development. Webview is useful when you want more customizations in the UI which is not possible using the native VS code API

**Comment by theReynald** [comment, marked as answer]
Hi Umar-shabbir,

I'm not an expert in this area but after a bit some research, seems that achieving HMR with VS Code Webview when developing a React-based extension can be a bit tricky, but it's possible.

Here's a general approach for achieving this:

**1. Setting up Webpack:**

  You'd typically use Webpack with the **webpack-dev-server** for HMR in regular web development. For a VS Code extension, you'd still use Webpack but handle the update mechanism a bit differently.
  
  Ensure you've set up Webpack with the necessary plugins:
  
  - react-hot-loader
  - webpack.HotModuleReplacementPlugin

**2. React Changes:**

- Wrap your app's main component with **`hot`** from **`react-hot-loader/root`**.

<img width="351" alt="Screenshot 2023-08-27 at 6 19 25 AM" src="https://github.com/microsoft/vscode-discussions/assets/100732140/a497a360-1dba-4c2a-96f2-4e91721edc0f">

**3. Webpack Configuration:**

Make sure your Webpack config has HMR enabled:
 
<img width="399" alt="Screenshot 2023-08-27 at 6 22 54 AM" src="https://github.com/microsoft/vscode-discussions/assets/100732140/2a43d02c-7eea-4fdb-b81f-8e00533d5e35">

**4. VS Code Webview Integration:**

You'll need to communicate between the Webview and the extension to notify the Webview when a new bundle is available.

- In your extension code, set up a watcher on your Webpack output directory. When changes are detected, send a message to the Webview to fetch the new bundle.

- In your Webview code, listen for the message from the extension. When you receive this message, re-fetch the bundle (or force a reload of the Webview's content).

**5. WebSocket Proxy:**

The VS Code Webview restricts many things for security, including WebSocket connections which are used by HMR. You may need to use or implement a proxy that communicates with the Webview via the allowed VS Code message passing API and forwards these messages to the actual WebSocket server.

**6. Reload Webview:**

One of the simpler ways to refresh the content is just to reload the entire Webview content. This won't give you the full HMR experience (keeping the state), but it will automatically show you the updated UI without a manual reload.

To achieve this, once you detect changes (via watching the Webpack output or some other mechanism), send a command to the Webview to run **location.reload()**.

The combination of these steps should provide you with an HMR-like experience within a VS Code Webview. It might not be as seamless as regular web development, but it should help to some extent.

**Reply by jiang-zhong-xi** [reply]
I resolve the same question by config webpack:
`
devServerConfig.client.webSocketURL = 'ws://127.0.0.1:3000/ws'
`

**Comment by tedjames** [comment]
Noticed there wasn't a good example of this anywhere. Here is my implementation of this with full page reload:

``` javascript
import * as vscode from "vscode";

export function activate(context: vscode.ExtensionContext) {
  let webview = vscode.commands.registerCommand("yourExtensionName.start", () => {
    let panel = vscode.window.createWebviewPanel(
      "webview",
      "React",
      vscode.ViewColumn.One,
      {
        enableScripts: true,
      }
    );

    // web is for my react root directory, rename for yours

    let scriptSrc = panel.webview.asWebviewUri(
      vscode.Uri.joinPath(context.extensionUri, "web", "dist", "index.js")
    );

    let cssSrc = panel.webview.asWebviewUri(
      vscode.Uri.joinPath(context.extensionUri, "web", "dist", "index.css")
    );

    // Watch the dist directory for changes
    const watcher = vscode.workspace.createFileSystemWatcher(
      new vscode.RelativePattern(context.extensionUri, "web/dist/*")
    );
    vscode.commands.executeCommand(
      "workbench.action.webview.reloadWebviewAction"
    );
    // React to changes in the dist directory
    watcher.onDidChange(() => {
      vscode.commands.executeCommand(
        "workbench.action.webview.reloadWebviewAction"
      );
    });
    watcher.onDidCreate(() => {
      vscode.commands.executeCommand(
        "workbench.action.webview.reloadWebviewAction"
      );
    });
    watcher.onDidDelete(() => {
      vscode.commands.executeCommand(
        "workbench.action.webview.reloadWebviewAction"
      );
    });

    panel.webview.html = `<!DOCTYPE html>
        <html lang="en">
          <head>
            <link rel="stylesheet" href="${cssSrc}" />
          </head>
          <body>
            <noscript>You need to enable JavaScript to run this app.</noscript>
            <div id="root"></div>
            <script src="${scriptSrc}"></script>
          </body>
        </html>
        `;
  });

  context.subscriptions.push(webview);
}

export function deactivate() {}
```

As you can see, no need for a JS event listener script or a window.location.reload(). We simple set up a few watchers to monitor for changes and then on change, execute the built-in vs code command as so:

```
  vscode.commands.executeCommand(
      "workbench.action.webview.reloadWebviewAction"
    );
```
